lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,361 @@
+ * ));
+ *
+ * You can also pass 'driverOptions' and any other documented option to each of this drivers to pass additional information.
+ */
+class PrimaryReplicaConnection extends Connection
+{
+    /** @var DriverConnection|null */
+    private $primary;
+
+    /**
+     * One replica picked at random
+     *
+     * @var DriverConnection|null
+     */
+    private $replica;
+
+    /**
+     * Creates Primary Replica Connection.
+     *
+     * @param mixed[] $params
+     *
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        array $params,
+        Driver $driver,
+        ?Configuration $config = null,
+        ?EventManager $eventManager = null
+    ) {
+        if (! isset($params['replica'], $params['primary'])) {
+            throw new InvalidArgumentException('primary or replica configuration missing');
+        }
+
+        if (count($params['replica']) === 0) {
+            throw new InvalidArgumentException('You have to configure at least one replica.');
+        }
+
+        $params['primary']['driver'] = $params['driver'];
+        foreach ($params['replica'] as $replicaKey => $replica) {
+            $params['replica'][$replicaKey]['driver'] = $params['driver'];
+        }
+
+
+        parent::__construct($params, $driver, $config, $eventManager);
+    }
+
+    private function switchToPrimary(): bool
+    {
+        $forcePrimaryAsReplica = $this->getTransactionNestingLevel() > 0;
+
+        if (isset($this->primary)) {
+            $this->_conn = $this->primary;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        $this->primary = $this->_conn = $this->connectTo('primary');
+
+        // Set replica connection to primary to avoid invalid reads
+        $this->replica = $this->primary;
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    public function connect(): bool
+    {
+        $connectionName = 'replica';
+
+        if ($this->_conn !== null) {
+            return false;
+        }
+
+        $forcePrimaryAsReplica = false;
+
+        if ($this->getTransactionNestingLevel() > 0) {
+            $connectionName        = 'primary';
+            $forcePrimaryAsReplica = true;
+        }
+
+        if (isset($this->$connectionName)) {
+            $this->_conn = $this->$connectionName;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        if ($connectionName === 'primary') {
+            $this->primary = $this->_conn = $this->connectTo($connectionName);
+
+            // Set replica connection to primary to avoid invalid reads
+            $this->replica = $this->primary;
```

Comment:
This is insane. Why instead of just keeping two connections separate and independent on the state, we assign one to the other? From the code readability standpoint, this is somewhat like "Monday = Tuesday".

Reply:
Should we just `unset($this->replica)`?